Flowdock notifications

t0mk opened the pull request:
Hi, I wrote a Flowdock notifier.

I did it a bit different way than the HipChat and Slack notifiers, I didn't want to duplicate code in e.g. serializing to JSON, and posting to HTTP URL.

There is quite a lot of duplicated code in manager/notifier.go. I was thinking I could rewite it in similar manner in another pull-request. Would you be interested in that?

I didn't write any tests, it wasn't straightforward to mock the config and the alert. Maybe it would be easier after the refactoring I suggest.

juliusv commented on config/config.proto:
```diff
@@ -70,6 +70,17 @@ message SlackConfig {
 
+message FlowdockConfig {
+  // Flowdock flow API token
+  optional string api_token = 2;
```
Heh. Please start numbers at 1 and then increment by 1 for each field :)

juliusv commented on config/config.proto:
```diff
@@ -70,6 +70,17 @@ message SlackConfig {
+  // Flowdock from_address
+  optional string from_address = 3;
+  // Flowdock flow tags
+  repeated string tags = 4;
```
You'll probably want to call this just `tag`, because when using a repeated field, you just list it multiple times with as the same field name:

```
tag: "foo"
tag: "bar"
...
```

juliusv commented on config/config.proto:
```diff
@@ -70,6 +70,17 @@ message SlackConfig {
+  optional string from_address = 3;
+  // Flowdock flow tags
+  repeated string tags = 4;
+  // Color of message when triggered.
```
The comment doesn't match the field below here...

juliusv commented on manager/notifier.go:
```diff
@@ -325,6 +326,59 @@ func (n *notifier) sendSlackNotification(op notificationOp, config *pb.SlackConf
+// compulsory fields in Flowdock JSON: source, from_address, subject, content
+// Content-type: application/json
+// POST to https://api.flowdock.com/v1/messages/team_inbox/:flow_api_token
+type FlowdockMessage struct {
```
AFAICS this type doesn't need to be exported / capitalized (but the fields need to be, for JSON unmarshaling).

juliusv commented on manager/notifier.go:
```diff
@@ -325,6 +326,59 @@ func (n *notifier) sendSlackNotification(op notificationOp, config *pb.SlackConf
+	return msg
+}
+
+func postJSONtoURL(jsonMessage []byte, url string) *http.Response {
```
I'd just call it `postJSON` (avoids the lower-case `to` awkardness etc.)

juliusv commented on manager/notifier.go:
```diff
@@ -325,6 +326,59 @@ func (n *notifier) sendSlackNotification(op notificationOp, config *pb.SlackConf
+	}
+	return buf
+}
+func getFlowdockNotificationMessage(op notificationOp, config *pb.FlowdockConfig, a *Alert) *FlowdockMessage {
```
`get...` -> `new...` seems to be the naming convention in Go for functions which just build objects for you.

juliusv commented on manager/notifier.go:
```diff
@@ -325,6 +326,59 @@ func (n *notifier) sendSlackNotification(op notificationOp, config *pb.SlackConf
+	Tags        []string `json:"tags,omitempty"`
+}
+
+func jsonize(msg interface{}) []byte {
```
`jsonize` sounds a bit... colloquial. How about `marshalJSON`?

juliusv commented on manager/notifier.go:
```diff
@@ -325,6 +326,59 @@ func (n *notifier) sendSlackNotification(op notificationOp, config *pb.SlackConf
+func jsonize(msg interface{}) []byte {
+	buf, err := json.Marshal(msg)
+	if err != nil {
+		glog.Errorln("Error compiling JSON:", err)
```
compiling -> marshalling or serializing would be the best term :)

juliusv commented:
Looks pretty cool otherwise.

It would be great indeed to refactor things more in the future and make tests easier.

t0mk commented:
I fixed issues from your comments. Plus,
- I added status (firing|resolved} as a tag for a Flowdock message.
- I mentioned Flowdock in readme
- I improved HTTP response logging in notifier.go

Once you merge this, will you rebuild the prom/alertmanager Docker image?

fabxc commented on manager/notifier.go:
```diff
@@ -494,6 +563,15 @@ func (n *notifier) handleNotification(a *Alert, op notificationOp, config *pb.No
+		}
+		flowdockMessage := newFlowdockNotificationMessage(op, fdConfig, a)
+		url := *flowdockURL + "/" + fdConfig.GetApiToken()
+		httpResponse := postJSON(marshallJSON(flowdockMessage), url)
```
Why would we want to make the request with invalid data?
I don't think the the `marshallJSON` helper is beneficial in general. It just hides the error from the control flow.

juliusv replied:
Oh yeah, I didn't even notice that. I agree... also, the name would be `marshalJSON` (with one `l`) - see also `MarshalJSON` in the stdlib.

juliusv commented on config/config.proto:
```diff
@@ -70,6 +70,17 @@ message SlackConfig {
 }
 
+message FlowdockConfig {
+  // Flowdock flow API token
```
Sorry if I didn't notice this before, but style nit: period after each doc comment please.

t0mk commented:
I fixed issues from previous comments. Also, I made the flowdock code look a bit more like the existing. I tried to put the code which I think can be reused in separate functions.

juliusv commented:
@t0mk Nice, thanks!
